[WIP] Remove BasicMeta/ordering of classes

asmeurer opened the pull request:
It has come to my attention that the ordering of classes (sympy/core/core.py) is causing some serious slowdowns in Python 2, due to a change in https://github.com/sympy/sympy/pull/12945 (doing `type(self) == type(other)` in `Basic.__eq__`, which calls `type(self).__cmp__(type(other))` in Python 2).  See https://github.com/sympy/sympy/issues/12898#issuecomment-317894206.

So here is my attempt at removing it (to fix https://github.com/sympy/sympy/issues/4269). This is ongoing work (I'll remove the "WIP" when it's done).

It's probably not going to be easy, so we'll see how far I actually get. We also need to pay careful attention to performance here. In particular, we need to replace the sorting in `Add` and `Mul` with something else, and whatever it is, we need to make sure it isn't slower (I'm first going to try `default_sort_key` and see how that goes).

If anyone wants to help out, feel free.

asmeurer committed:
Remove type sorting in meijerint _mytype

Sortability of SymPy types is going to be removed. The stated purpose of the
object is to create a hash representing the type of f. For general objects,
the args cannot be sorted anyway. For commutative objects, like Add and Mul,
the args are already sorted, and the sort key is based on the type (this could
change, but for now it is). So the sorting seems unnecessary.

asmeurer commented:
The first bug: whatever I did to meijerg broke this

```
>>> Integral(log(n)/n**3, (n, 1, oo)).doit()
meijerg(((-1, -2), (0,)), ((-1, -1, 0), ()), 1)/2
```

(the answer should be 1/4). Also, `(meijerg(((-1, -2), (0,)), ((-1, -1, 0), ()), 1)/2).is_number` incorrectly gives False.

asmeurer commented on sympy/integrals/meijerint.py:
```diff
@@ -293,7 +293,6 @@ def _mytype(f, x):
         res = []
         for t in types:
             res += list(t)
-        res.sort()
```
If anyone understands this function better and has a better suggestion, let me know (see the commit message). Basically, we can't sort `res` because `res` is a list of classes, and ordering of classes is what is being removed here. 

The whole function itself seems a bit shaky, and indeed, all the various ways I've tried to change it result in completely different (but I think mathematically equivalent) results from meijerint.

asmeurer replied:
Apparently it does need to be sorted. It's just a quick lookup, which is then guarded better with real matching. But if you don't sort, then something like `(Heaviside, log)` won't match `(log, Heaviside)`. I suppose we could also use frozensets.

asmeurer commented:
Is the meijerg answer wrong, or is evalf wrong?

```
>>> Integral(log(n)/n**3, (n, 1, oo)).doit()
meijerg(((-1, -2), (0,)), ((-1, -1, 0), ()), 1)/2
>>> Integral(log(n)/n**3, (n, 1, oo)).doit().evalf()
nan
```

smichr commented:
I recommend using ordered. I know you have expressed concerns regarding speed but I think they are unfounded. What would be a good benchmark for making a decision?

asmeurer commented:
I'm still trying to work out how to benchmark it. I'm hoping some of the benchmarks in https://github.com/sympy/sympy_benchmarks are sufficient.

The lines I'm referring to by the way are https://github.com/sympy/sympy/blob/b3f6a2808e18ce227faa5b950bbfa65eaceafc44/sympy/core/add.py#L16 and https://github.com/sympy/sympy/blob/b3f6a2808e18ce227faa5b950bbfa65eaceafc44/sympy/core/mul.py#L30 (basically, Basic.compare should go away).

asmeurer commented:
The is_number thing is https://github.com/sympy/sympy/issues/13071. 

About `meijerg(((-1, -2), (0,)), ((-1, -1, 0), ()), 1)`, I'm thinking mpmath is right, and it doesn't converge. It has a hard time with this I think because of the redundant 0's. Is it correct to remove them, as such: `meijerg(((-1, -2), ()), ((-1, -1), ()), 1)`?

If so, then it's defined in the limit

```
>>> meijerg(((-1, -2), ()), ((-1, -1), ()), 1).evalf()
nan
>>> meijerg(((-1, -2), ()), ((-1, -1), ()), 0.99999).evalf()
0.500008333441668
>>> meijerg(((-1, -2), ()), ((-1, -1), ()), 1.0000001).evalf()
0.499999916666677 - 2.59885244141122e-113*I
```

If you try the same thing with the 0's, mpmath just hangs.

asmeurer commented:
So my change was breaking the lookup in the meijerg algorithm, which made it fallback to a worse algorithm (which resulted in the meijerg).  I'll work on a better solution.

asmeurer committed:
Use frozenset in meijerint _mytype

The previous behavior of sorting needed to change (sorting of types is going
away). However, not sorting is also wrong, as we need the keys to be
canonicalized. Since the order does not appear to matter (the "type" is just
used as a quick lookup before doing a full match), just use a frozen set.

asmeurer committed:
Merge branch 'master' into rm-cmp

asmeurer commented:
Preliminary tests are inconclusive with `default_sort_key` / `ordered` vs. `Basic.compare`.

If I create a random list of symbols and create an Add or Mul out of it

```
s = list(symbols('x:1000'))
random.shuffle(s)
Mul(*s)
```

Then `default_sort_key` / `ordered` / `Basic.compare` run in about 40 / 50 / 95 ms. 

However, if I test `expand((x + y + z + 1)**32)`, I get 9.5 / 8.5 / 7.5 s.

This is all timed with the cache off. 

So I need a good real-life benchmark, that also results in a lot of nontrivial arg sorting in Add and/or Mul.

asmeurer commented:
Also some baseline benchmarks

No ordering: 15 ms and 4.5 s, resp.
Sorting with `id`: 15 ms and 4.8 s (very little difference from no sorting)

asmeurer commented:
OK, how about this?

```py
def fast_sort_key(x):
    if not x.args:
        return (type(x).__name__,) + x._hashable_content()
    return (type(x).__name__,) + tuple(fast_sort_key(i) for i in x.args)
```

It appears to be faster than both Basic.compare and default_sort_key in both the above cases.  It does a lot less work than default_sort_key. 

Some potential issues:

- Still uses the class name for sorting, so two different classes with the same name could confuse it. But: this is how things work now with Basic.compare, and you can always work around it by putting something in _hashable_content.

- Relies on _hashable_content being sortable. But the core only requires it to be a tuple. It could contain something unsortable like a frozenset.  So maybe we should create `fast_sort_key` which defaults to `_hashable_content` but which needs to be overridden in such cases.

- It doesn't try to handle equal items like `ordered`. I think this isn't an issue, because any two objects with the same `_hashable_content` should be equal. But one could see someone perhaps being lazy with _hashable_content and that not being the case (strictly speaking, only the reverse need hold, that equal implies same hashable_content).

Maybe this is abusing hashable_content for what it wasn't intended to do.

asmeurer commented:
fast_sort_key: 21 ms & 5 s on the above benchmarks

asmeurer committed:
Use default_sort_key instead of Basic.compare in Add and Mul

This is not completed yet (there are still test failures). I am not sure if
this is really the sort key we want to use. I am making this commit to make
benchmarking with asv easier.

asmeurer commented:
OK, I ran the sympy_benchmarks with asv on this branch: http://www.asmeurer.com/sympy_benchmarks/. The most recent commit (the right-most point there), is the move from Basic.compare to default_sort_key. So with those benchmarks, there isn't a noticeable difference. I might try to add some "core" benchmarks that can more accurately assess the performance of the core.

asmeurer commented:
I also ran the `asv continuous master rm-cmp` command

```
    before     after       ratio
  [0b61acbe] [fc5461f9]
+  849.00μs     1.04ms      1.23  solve.TimeMatrixSolve2.time_lusolve
+  879.90ms      1.01s      1.14  solve.TimeMatrixSolve.time_solve('GE')
+   42.72ms    48.58ms      1.14  solve.TimeMatrixOperations.time_det(4, 5)
+   52.72ms    59.73ms      1.13  large_exprs.TimeLargeExpressionOperations.time_count_ops
+  194.88μs   220.28μs      1.13  solve.TimeMatrixOperations.time_det_bareiss(3, 0)
+   55.91μs    62.31μs      1.11  solve.TimeMatrixArithmetic.time_dense_add(4, 5)
+    1.92ms     2.12ms      1.11  solve.TimeMatrixOperations.time_det_berkowitz(4, 2)
-  209.14ms   188.29ms      0.90  integrate.TimeIntegration01.time_doit
-  150.54ms   135.46ms      0.90  solve.TimeMatrixOperations.time_rank(4, 5)
-   47.46ms    42.21ms      0.89  solve.TimeMatrixOperations.time_det_bareiss(4, 5)
-     1.13s      1.00s      0.89  solve.TimeSolve01.time_solve
-   30.74ms    27.06ms      0.88  sum.TimeSum.time_doit
-  210.93μs   184.09μs      0.87  solve.TimeMatrixOperations.time_det(3, 0)
-  113.96ms    98.99ms      0.87  integrate.TimeIntegration01.time_doit_meijerg
-     1.05s   905.17ms      0.86  solve.TimeMatrixSolve.time_solve('LU')
-  214.90ms   183.72ms      0.85  large_exprs.TimeLargeExpressionOperations.time_jacobian_wrt_functions
-  929.53μs   793.14μs      0.85  solve.TimeMatrixArithmetic.time_dense_multiply(6, 0)
-    3.59ms     2.96ms      0.82  cse.TimeCSE.time_cse((4, 8))
-  531.83ms   384.37ms      0.72  solve.TimeMatrixSolve.time_solve('ADJ')
-    4.60ms     2.99ms      0.65  cse.TimeCSE.time_cse((3, 8))
-    5.33ms     2.79ms      0.52  cse.TimeCSE.time_cse((2, 8))

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
```

(left is master, right is this branch with default_sort_key).

It's a grab bag, it seems, but I think things are mostly getting faster (remember that x0.5 is the same as x2; I wish the values were reported log relative differences). It's hard to tell what is statistically significant, though.

asmeurer commented:
Well I'm giving up on telling if there's a real difference or not. I'm just going to stick with default_sort_key for now. Anything else I could come up with would be about the same as it anyway, so if we end up discovering it is slow, we just need to figure out how to make it faster.

asmeurer committed:
Remove Basic.compare and 'old' print ordering

Also removes everything that depends on them. I haven't yet fixed all test
failures caused by this, or the previous commit.

The old order is heavily tied to Basic.compare and the ordering of classes. I
don't see a value in keeping it with a deprecation, as it would require also
keeping the ordering of classes. The name implies that it is de facto
deprecated, and it has been called 'old' for some time now. The main use-case
was supposedly that 'old' was faster, but I'm not even clear that that should
be the case (it still does sorting), and even if it is, if you care about
speed, you should use order='none'.

I don't think Basic.compare itself should be kept deprecated either. All uses
of it should use default_sort_key. The ordering will not be equivalent (which
is one reason why I don't think a deprecation would be that helpful), but the
ordering here is arbitrary anyway.

asmeurer committed:
Remove BasicMeta

The class definition is still there (I will delete it in a future commit). I
haven't fixed any test failures yet, just made it so that sympy can be
imported.

asmeurer commented:
OK, so default_sort_key does have some real performance issues. The problem is that it uses `as_ordered_terms`, which (through `as_terms`), calls [evalf](https://github.com/sympy/sympy/blob/8ff744d843f9dc78c560e61e3edc01f878d1960b/sympy/core/expr.py#L983). Unfortunately, evalf is not always fast. It can hang on things like summations and integrals. Even ignoring the hanging, it's not fast enough that it should be called constantly in the core. 

I imagine that code is there to give a nice numeric ordering for printing, but for Add and Mul args, we don't need a nice ordering, just a consistent one.

asmeurer commented:
Uncovering a lot of untested issues in master with this. Maybe we should randomize the sort key every once in awhile to make sure everything is tested.

asmeurer committed:
Don't use as_ordered_terms/factors in sort_key

These functions call evalf to order terms numerically, which is too slow for a
sort key. The sort key is used in the core Add and Mul ordering, among other
places, where we don't care about numeric ordering. We just want something
consistent. But it does need to be as fast as possible.

asmeurer committed:
Fix issubclass call in printer.py

asmeurer committed:
Fix importing of meijerint_doc

asmeurer committed:
Don't test string input in test_args

asmeurer committed:
Fix sympify -> _sympify in sympy/core/containers.py

asmeurer committed:
Don't use Tuple to sympify in verify_numerically

Now that Tuple uses _sympify instead of sympify, it cannot contain None.

asmeurer committed:
Revert disallowing strings in Tuple

It's too much of a backwards compatibility break, since it also breaks
sympify(('1', '2')), and I don't think it's actually needed to fix what I
originally thought it was needed for.

asmeurer committed:
Don't assume the order of unevaluated expressions in as_independent test

asmeurer commented:
@smichr I'm trying to understand this test:

```py

def test_AssocOp_Function():
    e = S('''
    Min(-sqrt(3)*cos(pi/18)/6 + re(1/((-1/2 - sqrt(3)*I/2)*(1/6 +
    sqrt(3)*I/18)**(1/3)))/3 + sin(pi/18)/2 + 2 + I*(-cos(pi/18)/2 -
    sqrt(3)*sin(pi/18)/6 + im(1/((-1/2 - sqrt(3)*I/2)*(1/6 +
    sqrt(3)*I/18)**(1/3)))/3), re(1/((-1/2 + sqrt(3)*I/2)*(1/6 +
    sqrt(3)*I/18)**(1/3)))/3 - sqrt(3)*cos(pi/18)/6 - sin(pi/18)/2 + 2 +
    I*(im(1/((-1/2 + sqrt(3)*I/2)*(1/6 + sqrt(3)*I/18)**(1/3)))/3 -
    sqrt(3)*sin(pi/18)/6 + cos(pi/18)/2))''')
    # the following should not raise a recursion error; it
    # should raise a value error because the first arg computes
    # a non-comparable (prec=1) imaginary part
    raises(ValueError, lambda: e._eval_evalf(2))
```

Something in this branch is changing it so that `e._eval_evalf(2)` gives `0e+0` instead of raising ValueError.  Should that still be considered a correct result? My understanding is that these "0e0" numbers mean evalf couldn't determine the quantity. 

What should the true value of that constant be? `e.simplify()` in this branch gives `-sqrt(3)*cos(pi/18)/3 - sin(pi/18) + 2`, which evalfs to `1.25777280103144`. Is that correct? Or is simplify itself being fooled by evalf?

asmeurer committed:
Fix order of string in test

asmeurer committed:
Merge branch 'rm-cmp' of github.com:asmeurer/sympy into rm-cmp

asmeurer committed:
Fix some printing tests due to different ordering

asmeurer committed:
Merge branch 'master' into rm-cmp

asmeurer committed:
Fix some doctests for different ordering of printing

The diophantine test appears to be different, but it's mathematically
equivalent to a change of variables t -> -t.

asmeurer committed:
Fix changed output of a integral transform doctest

The two expressions appear different, but I believe they are equivalent, at
least for sufficient assumptions on a and k (k should be implicitly positive
as it is the variable of the Hankel transform).

asmeurer commented:
Latest issue:

```py
>>> collect(x*Derivative(f(x), x) - 6*x + f(x)*Derivative(f(x), x) + f(x), f(x).diff(x))
x*Derivative(f(x), x) - 6*x + (Derivative(f(x), x) + 1)*f(x)
```

This is causing a match in the ODE module to fail.

asmeurer commented:
I think collect is implicitly relying on the fact that `(f(x)*diff(f(x), x)).args` gives the derivative first.

asmeurer committed:
Fix a doctest for changed printing order

asmeurer commented on sympy/integrals/transforms.py:
```diff
@@ -1799,7 +1799,7 @@ def hankel_transform(f, r, k, nu, **hints):
     >>> ht = hankel_transform(exp(-a*r), r, k, 0)
     >>> ht
-    a/(k**3*(a**2/k**2 + 1)**(3/2))
+    1/(a**2*(1 + k**2/a**2)**(3/2))
```
Can anyone confirm these are the same? In both master and this branch, `noconds=False` gives the conditions `Eq(Abs(periodic_argument(polar_lift(k)**2, oo)), 0) & (Abs(periodic_argument(polar_lift(a)**2, oo)) < pi)`.

asmeurer replied:
I think that at least means k is positive (which I already guessed based on the integration limits of the inverse transform). The `a` condition I'm still tring to figure out (c.f. https://github.com/sympy/sympy/issues/11166). Does it mean `a` is in the first quadrant? Do these conditions allow you to modify the square roots in a way that reduces one form to the other?

jksuom replied:
The two expressions agree for positive `a` (and `k`) with `a/(a**2 + k**2)**(3/2)`. The  `a`  condition means that the argument of `a` is between `-pi/2` and `pi/2` so that `Re a > 0`. Then `a**2`, `a**2 /k**2 + 1` and `1 + k**2/a**2` do not lie on the negative real axis which is the branch cut of `z**(3/2)`, and therefore both expressions define continuous analytic functions of  `a` in `Re a > 0`.

Since the positive half-plane is connected it follows from the principle of analytic continuation that the two functions of `a` agree for `Re a > 0`.

czgdp1807 commented:
@asmeurer Please provide update on this PR. Do we still need this one?

asmeurer commented:
We still need to do this, but it's a huge undertaking, because it changes the arg ordering for Add and Mul. It might be a little easier once we drop Python 2. This work will probably need to be restarted. My work here can be reused in parts, but likely a lot will have to be done from scratch. I don't know when I would have the time to work on it again, so if someone else wants to give it a go, please do.

oscarbenjamin commented:
I don't think this should have an attached release milestone.

asmeurer commented:
No it shouldn't. I already worked around the performance issue in Python 2 in a separate pull request.

czgdp1807 commented:
Closing in favour of https://github.com/sympy/sympy/pull/19907
